### ISSUE TITLE
May i ask when you will upload the training code and pre trained model approximately?

### ISSUE THREAD

**Issue by Janne-byti** [polimi-ispl/PARCnet#2]
Your work is excellent. 
I have read your paper and it mentioned that the algorithm's performance exceeds all baseline models, and the number of model parameters is very small.
 I would like to ask if you have tested the algorithm's performance on a voice dataset other than music? 
Have you also tested the real-time running efficiency of algorithms on the computer? I have measured that the LPCNet algorithm has an RTF of 3 and cannot run in real-time. Additionally, the local RTF of the TFGAN algorithm is around 0.6 (which is consistent with the data provided in the author's paper). 
According to the paper, should your algorithm have a real-time rate that is half of that of the TFGAN algorithm?

**Comment by EMMALIKECODE123** [comment]
> Your work is excellent. I have read your paper and it mentioned that the algorithm's performance exceeds all baseline models, and the number of model parameters is very small. I would like to ask if you have tested the algorithm's performance on a voice dataset other than music? Have you also tested the real-time running efficiency of algorithms on the computer? I have measured that the LPCNet algorithm has an RTF of 3 and cannot run in real-time. Additionally, the local RTF of the TFGAN algorithm is around 0.6 (which is consistent with the data provided in the author's paper). According to the paper, should your algorithm have a real-time rate that is half of that of the TFGAN algorithm?

hello janne,
I am working on a PLC for voice too, and I want to run other person's pre-model first, I saw you have run LPCNET and TFGAN model, so can you give me a link of LPCNET and TFGAN, Thank you so much。
finally, i have found the FARCNET pre-model which can run with command
git checkout a454733399577c527e584978d35c17547542686e
hope this can help you.

**Comment by Janne-byti** [comment]
@EMMALIKECODE123 
I am happy to discuss the implementation and optimization of PLC algorithm with you. Can you provide me with an email address or other ways to contact you？

**Comment by EMMALIKECODE123** [comment]
> @EMMALIKECODE123 I am happy to discuss the implementation and optimization of PLC algorithm with you. Can you provide me with an email address or other ways to contact you？

yongyaqin1225@126.com

**Comment by ilic-mezza** [comment]
Hi @Janne-byti,

Thank you very much for the interest! 

First, I have fixed the [issue](https://github.com/polimi-ispl/PARCnet/issues/1#issuecomment-1870901480) with the corrupted pretrained model (which was supposed to be always available and about 5 MB). 

Second, our training code is heavily under-optimized at the moment. It preprocesses and saves packets in an offline mode rather than loading and chunking audio files at training time. This is mainly due to the fact that PARCnet was developed with an early implementation of the linear predictor which was so slow that consituted a bottleneck during training. 

We are thinking to refactor the training code before releasing it. At the same time, we are also considering to release it as-is, and update it at a later date. Either way, some version of the training code will soon be available.

As of now, we have not tried PARCnet on a speech dataset. PARCnet's linear branch performs LPC so it might prove to be a good fit for concealing speech packets too. 

In [our paper](https://ieeexplore.ieee.org/document/10360264), we tested the real-time ratio of all the models on a laptop CPU, including LPCNet and [TFGAN](https://github.com/Guanyuansheng/TFGAN-PLC). See the last column of Table 2. 

Although we re-trained the PLC version of LPCNet using the Python/Keras implementation, we tested using the efficient implementation written in C that can be found on the `plc_challenge` branch of the LPCNet repo ([https://github.com/xiph/LPCNet/tree/plc_challenge](https://github.com/xiph/LPCNet/tree/plc_challenge)). This allows LPCNet to run in real-time on our machine (AMD Ryzen 5900HS).

Moreover, we report the CPU time of PARCnet considering the neural network forward pass only. We did not implement the parallelization of the two branches yet. In our code, linear and nonlinear predictions take place sequentially. The actual RTR is thus expected to be higher than what reported. However, with a little bit of work, the two predictors could be run in parallel, so that the total execution time is dominated by the slowest model, i.e., the neural network.

Our neural network will be faster than TFGAN (not sure about x2 faster, but still...) After all, our neural network is a stripped-down version of TFGAN with nearly a quarter of the parameters.

**Comment by Janne-byti** [comment]
@ilic-mezza 
Thank you for your reply!
I tried using your pre trained model to handle voice packet loss, and in fact, it also had some effect. I also tested the real-time rate of the model, which is about half of TFGAN. But the speech situation may be more complex, such as noise and other issues, so I really want to retrain on the speech dataset as soon as possible to verify the performance of PARCNet in speech. This is a matter that requires a lot of time,but I am very interested in it.
Therefore, if possible, I hope you can release the original code version as soon as possible and gradually optimize it later. I will also create a voice branch and look forward to discussing it with you. Perhaps we can find new optimization directions.
